plugins/com.google.cloud.tools.eclipse.util/src/com/google/cloud/tools/eclipse/util/ArtifactRetriever.java

```diff
@@ -0,0 +1,196 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.eclipse.util;
+
+import com.google.cloud.tools.eclipse.util.status.StatusUtil;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableSortedSet;
+import com.google.common.collect.ImmutableSortedSet.Builder;
+import org.apache.maven.artifact.versioning.ArtifactVersion;
+import org.apache.maven.artifact.versioning.DefaultArtifactVersion;
+import org.apache.maven.artifact.versioning.VersionRange;
+import org.eclipse.core.runtime.CoreException;
+import org.w3c.dom.Document;
+import org.w3c.dom.NodeList;
+import org.xml.sax.SAXException;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.NavigableSet;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.xml.parsers.DocumentBuilderFactory;
+import javax.xml.parsers.ParserConfigurationException;
+import javax.xml.xpath.XPath;
+import javax.xml.xpath.XPathConstants;
+import javax.xml.xpath.XPathFactory;
+
+/**
+ * {@link ArtifactRetriever} provides access to Maven artifacts using low-level URL and XPath APIs
+ * rather than using the M2E plugin to work around shortcomings in the ability of M2E to query
+ * Maven for available versions. Additionally, M2E APIs are internal and unstable, and thus may
+ * change between versions.
+ *
+ * <p>The artifact retriever reads Maven Central metadata XML files to retrieve available and latest
+ * versions.
+ */
+public class ArtifactRetriever {
+  
+  private static final Logger logger = Logger.getLogger(ArtifactRetriever.class.getName());
+  
+  @VisibleForTesting
+  static URL getMetadataUrl(String groupId, String artifactId) {
+    String groupPath = groupId.replace('.', '/');
+    try {
+      return new URL(
+          "https://repo1.maven.org/maven2/" + groupPath + "/" + artifactId + "/maven-metadata.xml");
+    } catch (MalformedURLException e) {
+      throw new IllegalStateException(
+          String.format("Could not construct metadata URL for artifact %s", artifactId), e);
+    }
+  }
+
+  private final LoadingCache<String, ArtifactVersion> latestVersion =
+      CacheBuilder.newBuilder()
+          .refreshAfterWrite(4, TimeUnit.HOURS)
+          .build(
+              new CacheLoader<String, ArtifactVersion>() {
+
+                @Override
+                public ArtifactVersion load(String coordinates) throws Exception {
+                  Document document = getMetadataDocument(coordinates);
```

Comment:
It's the network fetch that dominates here, right?  So this seems pretty important to address.